src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -1586,6 +1543,115 @@ void MPS::propagate_to_neighbors_internal(uint_t min_qubit, uint_t max_qubit,
   }
 }
 
+reg_t MPS::sample_measure(const reg_t &qubits, const rvector_t &rnds) const {
+  reg_t internal_qubits = get_internal_qubits(qubits);
+  return sample_measure_internal(internal_qubits, rnds);
+}
+
+// The algorithm implemented here is based on https://arxiv.org/abs/1709.01662.
+// Given a particular base value, e.g., 11010, its probability is computed by contracting
+// the suitable matrices per qubit (from right to left), i.e., mat(0) for qubit 0, mat(1) 
+// for qubit 1, mat(0) for qubit 2, mat(1) for qubit 3, mat(1) for qubit 4.
+// We build the randomly selected base value for every shot as follows:
+// For the first qubit, compute its probability for 0 and then randomly select 
+//        the measurement. 'mat' is initialized to the suitable matrix (0 or 1).
+// For qubit i, we store in 'mat'(the contraction of the matrices that were selected up to i-1).
+//        We compute the probability that qubit i is 0 by contracting with matrix 0. 
+//        We randomly select a measurement according to this probability. We then update 'mat' 
+//        by contracting it with the suitable matrix (0 or 1).
+
+reg_t MPS::sample_measure_internal(const reg_t &qubits, const rvector_t &rnds) const {
+  uint_t size = qubits.size();
+  double prob = 1;
+  std::string current_measure="";
+  char measure_1_qubit;
+  bool is_first_qubit = true;
+  cmatrix_t mat;
+
+  for (uint_t i=0; i<size; i++) {
+    measure_1_qubit = sample_measure_single_qubit(i, is_first_qubit, current_measure, 
```

Comment:
Currently `i` is 0 iff `is_first_qubit` is `True`

Reply:
ok, changed to `if (qubit == 0)`